File: day12.py
#!/usr/bin/env python3

from get_input import get_input
from json import loads

def find_numbers(doc, ignore=None):
    ignore = ignore or []
    queue = [doc]
    total = 0
    while queue:
        item = queue.pop()
        if type(item) is int:
            total += item
        elif type(item) is dict:
            item_values = item.values()
            if not any(ignored in item_values for ignored in ignore):
                queue.extend(item_values)
        elif type(item) is list:
            queue.extend(item)
    return total

def part1(doc):
    return find_numbers(doc)

def part2(doc):
    return find_numbers(doc, ignore=['red'])

if __name__ == '__main__':
    json_doc = loads(get_input(day=12, year=2015))
    print("Part 1: {}".format(part1(json_doc)))
    print("Part 2: {}".format(part2(json_doc)))

File: day12.rs
use common::get_input;
use std::collections::VecDeque;
use serde_json::{self, Value};

type Input = Value;

fn part1(input: &Input) -> i64 { 
    let mut queue: VecDeque<&Value> = VecDeque::from(vec![input]);
    let mut total = 0;
    while let Some(item) = queue.pop_back() {
        match item {
            Value::Null => unimplemented!(),
            Value::Bool(_) => unimplemented!(),
            Value::Number(n) => total += n.as_i64().unwrap() as i64,
            Value::String(_) => {},
            Value::Array(v) => queue.append(&mut v.iter().collect()),
            Value::Object(m) => queue.append(&mut m.values().collect()),
        }
    };
    total
}

fn part2(input: &Input) -> i64 { 
    let mut queue: VecDeque<&Value> = VecDeque::from(vec![input]);
    let mut total = 0;
    while let Some(item) = queue.pop_back() {
        match item {
            Value::Null => unimplemented!(),
            Value::Bool(_) => unimplemented!(),
            Value::Number(n) => total += n.as_i64().unwrap() as i64,
            Value::String(_) => {},
            Value::Array(v) => queue.append(&mut v.iter().collect()),
            Value::Object(m) => {
                if !m.values().any(|v| v == "red") {
                    queue.append(&mut m.values().collect())
                }
            },
        }
    };
    total
}

fn parse(text: String) -> Input {
    serde_json::from_str(text.as_str()).unwrap()
}

fn main() {
    let input = parse(get_input(12, 2015));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}

